plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSProjectSetupValidator.xtend

```diff
@@ -664,14 +661,45 @@ class N4JSProjectSetupValidator extends AbstractN4JSDeclarativeValidator {
 					addIssue(getMessageForNON_EXISTING_PROJECT(id), it, NON_EXISTING_PROJECT);
 				} else {
 					// Create only one single validation issue for a particular project reference.
-					var valid = true;
-					if (valid && desc?.projectId == id) {
+					if (desc?.projectId == id) {
 						addProjectReferencesItselfIssue(it.eContainer, features.last, references.indexOf(it));
-						valid = false;
-					}
-					if (valid && !projectPredicate.apply(project)) {
+					}else if (!projectPredicate.apply(project)) {
 						addInvalidProjectTypeIssue(it.eContainer, id, project.projectType, features.last, references.indexOf(it));
-						valid = false;
+					}else{
+						if(it instanceof ProjectDependency){
```

Comment:
this looks like a good location to me to extract some logic into another method 😃